config/reflection_magic.go

```diff
@@ -97,11 +98,32 @@ func makeArgumentConstructors(fnType reflect.Type, argTypes map[reflect.Type]con
 	return out, nil
 }
 
+func checkConstructorOpts(t reflect.Type, opts ...interface{}) error {
+	if !t.IsVariadic() {
+		if len(opts) > 0 {
+			return errors.New("constructor doesn't accept any options")
+		}
+		return nil
+	}
+	// variadic parameters always go last
+	optType := t.In(t.NumIn() - 1).Elem()
+	for _, opt := range opts {
+		if opt == nil {
+			return errors.New("expected a transport option, got nil")
+		}
+		if reflect.TypeOf(opt) != optType {
```

Comment:
From [the spec](https://golang.org/ref/spec#Passing_arguments_to_..._parameters):

> Otherwise, the value passed is a new slice of type []T with a new underlying array whose successive elements are the actual arguments, which all must be assignable to T.

That is, checking for identical types seems wrong; you should instead use `opt.AssignableTo(wantType)`.

You should add a test case covering this, too. A fairly normal example would be for an Option to be of type `type Limit int64`, and for you to pass a regular old `int64` value. Not identical types, but assignable.

I get that this will likely not happen in practice, as I seem to understand practically all options will be function types. But the named-vs-unnamed distinction could also apply to functions, given that `type Option` is a named type and not an alias type. And I think it would make more sense to apply the same rules that the Go spec does for variadic arguments.

https://golang.org/ref/spec#Assignability has more details on the rules. Most rules seem like they'd never apply, so we don't need to worry about e.g. the line involving channels. You might want to special case untyped nil and reject it, though, as I think it never makes sense as an option.

Reply:
(ah, you already reject untyped nil for incoming options, good!)

Reply:
Good point. `AssignableTo` doesn't work though, it's `ConvertibleTo` that we're looking for here. We then also need to explicitly do the type conversion using `reflect.Value.Convert`.
Only caveat here: There's a potential failure case when the argument is for example an `int32`, and you pass a value larger than `math.MaxInt32` to the constructor. Go will happily convert this value, leading to an overflow. I'm not very concerned about this corner case in practice though.

Reply:
How did you arrive at the conclusion that you want convertible? I'm fairly sure you want assignable, which is what the spec says, and which does not have the problem with lossy conversions that you mention. Assignments are never lossy, because the rules are stricter than explicit conversions.

Reply:
A `type limit int` is not `AssignableTo` an `int`, but it is `ConvertibleTo` it.

Reply:
You're right, that was a bad example. That follows the rules we want, I think, because `int` is a named type. I feel like integers are a bit of a red herring here - in practice you're only going to care about function types, I'm pretty sure.

Reply:
I'm not sure I'm following. With functions, we'd only use the functions defined in the package that defined the `Option` anyway, so there's no need for converting or assigning anything?

Reply:
Right. For functions, checking for type identity via `==` is enough. I was just saying that, from a Go spec perspective, checking for assignability is more correct and consistent with what people might expect. In practice, I imagine either will be fine.